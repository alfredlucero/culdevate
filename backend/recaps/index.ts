import { NowRequest, NowResponse } from "@now/node";

const sampleRecaps = [
  {
    title: "Summaries and Goals",
    subtitle: "Subtitle: Position, Emphasis, Etc.",
    bulletPoints: ["I endeavor to achieve something!", "Software architect!"],
    type: "objectives",
    startDate: "2019/04/20",
    endDate: null,
  },
  {
    title: "Sandia National Labs",
    subtitle: "Subtitle: Position, Emphasis, Etc.",
    bulletPoints: ["Web Developer R&D Intern", "Software Development Technical Intern"],
    type: "work",
    startDate: "2016/06/20",
    endDate: "2017/06/10",
  },
  {
    title: "University of California, Los Angeles",
    subtitle: "BS in Computer Science, 3.56 GPA",
    bulletPoints: ["Upsilon Pi Epsilon Honor Society 2016-2017", "Daily Bruin Web Developer Intern"],
    type: "education",
    startDate: "2013/10/01",
    endDate: "2017/6/10",
  },
  {
    title: "SJB Tennis",
    subtitle: "Spring Season",
    bulletPoints: ["Varsity Tennis all 4 years", "Captain - Senior Year", "MVP - Third Year"],
    type: "extracurricular",
    startDate: "2009/09/15",
    endDate: "2013/5/20",
  },
  {
    title: "Terraform",
    subtitle: "Subtitle: Position, Emphasis, Etc.",
    bulletPoints: ["Learned how to do infrastructure as code with AWS stuff!"],
    type: "skills",
    startDate: "2018/01/10",
    endDate: null,
  },
  {
    title: "Promoted to Software Engineer II",
    subtitle: "Subtitle: Position, Emphasis, Etc.",
    bulletPoints: ["After 1 year and many projects, I earned my stripes!"],
    type: "achievements",
    startDate: "2018/10/20",
    endDate: null,
  },
  {
    title: "Arm in Arm Award",
    subtitle: "Subtitle: Position, Emphasis, Etc.",
    bulletPoints: ["Helped out CGE with Billing Rewrite to steer them back on time."],
    type: "awards",
    startDate: "2019/01/07",
    endDate: "2019/05/20",
  },
  {
    title: "Reference from Sandia Mentors",
    subtitle: "Subtitle: Position, Emphasis, Etc.",
    bulletPoints: ["Can check them out on LinkedIn"],
    type: "proof",
    startDate: "2016/06/20",
    endDate: "2017/06/05",
  },
  {
    title: "Code Complete Book",
    subtitle: "Subtitle: Position, Emphasis, Etc.",
    bulletPoints: ["Finished reading this book!"],
    type: "other",
    startDate: "2018/10/12",
    endDate: "2019/07/25",
  },
];

module.exports = async (req: NowRequest, res: NowResponse) => {
  console.log(req);

  const test = await Promise.resolve({ blah: "test" });

  const blah = { test: "mate",

};
console.log(blah);

  console.log(test);
  res.json(sampleRecaps);
};
