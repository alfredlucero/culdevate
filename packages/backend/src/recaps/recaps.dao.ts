import {
  RecapBaseModel,
  RecapWorkExperienceModel,
  RecapEducationModel,
  RecapAccomplishmentsModel,
  RecapPublicationsModel,
  RecapSkillsModel,
  RecapSideProjectsModel,
  RecapOrganizationsModel,
  RecapReferencesModel,
  RecapOtherModel,
  Recap,
} from "./recaps.model";

const RecapsDao = {
  findAllRecaps(userId: string) {
    return RecapBaseModel.find({ userId });
  },

  findRecapById(recapId: string) {
    return RecapBaseModel.findById(recapId);
  },

  findRecapByIdAndUserId({ recapId, userId }: { recapId: string; userId: string }) {
    return RecapBaseModel.findOne({ userId, _id: recapId });
  },

  createRecap(recap: Recap) {
    switch (recap.kind) {
      case "Work Experience":
        const workExperienceRecap = new RecapWorkExperienceModel(recap);
        return workExperienceRecap.save();
      case "Education":
        const educationRecap = new RecapEducationModel(recap);
        return educationRecap.save();
      case "Accomplishments":
        const accomplishmentsRecap = new RecapAccomplishmentsModel(recap);
        return accomplishmentsRecap.save();
      case "Publications":
        const publicationsRecap = new RecapPublicationsModel(recap);
        return publicationsRecap.save();
      case "Skills":
        const skillsRecap = new RecapSkillsModel(recap);
        return skillsRecap.save();
      case "Side Projects":
        const sideProjectsRecap = new RecapSideProjectsModel(recap);
        return sideProjectsRecap.save();
      case "Organizations":
        const organizationsRecap = new RecapOrganizationsModel(recap);
        return organizationsRecap.save();
      case "References":
        const referencesRecap = new RecapReferencesModel(recap);
        return referencesRecap.save();
      case "Other":
        const otherRecap = new RecapOtherModel(recap);
        return otherRecap.save();
      default:
        throw new Error("Failed to find matching recap kind to create!");
    }
  },

  updateRecapById({ recapId, updatedRecap }: { recapId: string; updatedRecap: Recap }) {
    switch (updatedRecap.kind) {
      case "Work Experience":
        return RecapWorkExperienceModel.findOneAndUpdate(
          { _id: recapId },
          { $set: updatedRecap, ...(!updatedRecap.endDate ? { $unset: { endDate: 1 } } : {}) },
          { new: true },
        );
      case "Education":
        return RecapEducationModel.findOneAndUpdate(
          { _id: recapId },
          { $set: updatedRecap, ...(!updatedRecap.endDate ? { $unset: { endDate: 1 } } : {}) },
          { new: true },
        );
      case "Accomplishments":
        return RecapAccomplishmentsModel.findOneAndUpdate({ _id: recapId }, updatedRecap, { new: true });
      case "Publications":
        return RecapPublicationsModel.findOneAndUpdate({ _id: recapId }, updatedRecap, { new: true });
      case "Skills":
        return RecapSkillsModel.findOneAndUpdate({ _id: recapId }, updatedRecap, { new: true });
      case "Side Projects":
        return RecapSideProjectsModel.findOneAndUpdate(
          { _id: recapId },
          { $set: updatedRecap, ...(!updatedRecap.endDate ? { $unset: { endDate: 1 } } : {}) },
          { new: true },
        );
      case "Organizations":
        return RecapOrganizationsModel.findOneAndUpdate(
          { _id: recapId },
          { $set: updatedRecap, ...(!updatedRecap.endDate ? { $unset: { endDate: 1 } } : {}) },
          { new: true },
        );
      case "References":
        return RecapReferencesModel.findOneAndUpdate({ _id: recapId }, updatedRecap, { new: true });
      case "Other":
        return RecapOtherModel.findOneAndUpdate({ _id: recapId }, updatedRecap, { new: true });
      default:
        throw new Error("Failed to find matching recap kind to update!");
    }
  },

  removeRecapById(recapId: string) {
    return RecapBaseModel.findByIdAndDelete({ _id: recapId });
  },
};

export default RecapsDao;
