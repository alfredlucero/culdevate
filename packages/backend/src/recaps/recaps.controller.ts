import { Response } from "express";
import { RequestWithUser } from "../interfaces/requestWithUser";
import RecapsDao from "./recaps.dao";
import { Recap } from "./recaps.model";

const RecapsController = {
  async createRecap(req: RequestWithUser, res: Response) {
    const currentUser = req.user;
    const recap: Recap = {
      ...req.body,
      userId: currentUser.id,
    };

    try {
      const createdRecap = await RecapsDao.createRecap(recap);

      res.status(201).json(createdRecap);
    } catch (err) {
      res.status(500).json({ message: "Failed to create recap." });
    }
  },

  async updateRecap(req: RequestWithUser, res: Response) {
    const currentUser = req.user;
    const recapId = req.params.recapId;
    const updatedRecap: Recap = {
      ...req.body,
      userId: currentUser.id,
    };

    try {
      const foundRecap = await RecapsDao.findRecapById(recapId);
      const isMatchingUserId = foundRecap && foundRecap.userId.toString() === currentUser.id;

      if (!foundRecap || !isMatchingUserId) {
        return res.status(404).json({ message: "Failed to find matching recap to update" });
      }

      const changedRecap = await RecapsDao.updateRecapById({ recapId, updatedRecap });

      if (changedRecap) {
        res.status(200).json(changedRecap);
      } else {
        // If by some chance it got deleted while we were trying to update it
        // we'll still get a 404 back
        res.status(404).json({ message: "Failed to find matching recap to update" });
      }
    } catch (err) {
      res.status(500).json({ message: "Failed to update recap" });
    }
  },

  async deleteRecap(req: RequestWithUser, res: Response) {
    const recapId = req.params.recapId;
    const currentUser = req.user;

    try {
      const foundRecap = await RecapsDao.findRecapById(recapId);
      const isMatchingUserId = foundRecap && foundRecap.userId.toString() === currentUser.id;

      if (!foundRecap || !isMatchingUserId) {
        return res.status(404).json({ message: "Failed to find matching recap to delete" });
      }

      const deletedRecap = await RecapsDao.removeRecapById(recapId);

      if (deletedRecap) {
        res.status(200).json(deletedRecap);
      } else {
        // If by some chance it got deleted while we were trying to delete it
        // we'll still get a 404 back
        res.status(404).json({ message: "Failed to find matching recap to delete" });
      }
    } catch (err) {
      res.status(500).json({ message: "Failed to delete recap" });
    }
  },
};

export default RecapsController;
