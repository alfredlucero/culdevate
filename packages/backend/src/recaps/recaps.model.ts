import mongoose from "mongoose";
import { MAX_BULLETPOINT_LENGTH, MAX_BULLETPOINTS, MAX_GENERAL_LENGTH } from "./recaps.validation";

export interface RecapBase {
  kind: RecapKind;
  startDate?: Date;
  endDate?: Date;
  bulletPoints: string[];
}

export type RecapKind =
  | "WorkExperience"
  | "Education"
  | "Accomplishments"
  | "Skills"
  | "SideProjects"
  | "Organizations"
  | "References"
  | "Other";

const recapBaseOptions = {
  discriminatorKey: "kind",
  collection: "recaps",
};

const recapBaseSchema = new mongoose.Schema(
  {
    startDate: { type: Date },
    endDate: { type: Date },
    bulletPoints: {
      type: [
        {
          type: String,
          maxlength: MAX_BULLETPOINT_LENGTH,
        },
      ],
      required: true,
      maxlength: MAX_BULLETPOINTS,
    },
  },
  recapBaseOptions,
);

export interface RecapBaseModel extends RecapBase, mongoose.Document {}

// Base Recap model from which other Recap models will inherit from based on the
// the value of the "kind" discriminator
export const RecapBaseModel = mongoose.model<RecapBaseModel>("Recap", recapBaseSchema);

// Work Experience

export interface RecapWorkExperience {
  title: string;
  location: string;
  company: string;
  employmentType: EmploymentType;
}

export type EmploymentType = "partTime" | "fullTime";

const recapWorkExperienceSchema = new mongoose.Schema({
  title: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  location: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  company: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  employmentType: { type: String, required: true, enum: ["partTime", "fullTime"] },
});

export interface RecapWorkExperienceModel extends RecapBaseModel, RecapWorkExperience {}

const RecapWorkExperience = RecapBaseModel.discriminator("WorkExperience", recapWorkExperienceSchema);

export const RecapWorkExperienceModel = mongoose.model<RecapWorkExperienceModel>("WorkExperience");

// Education

export interface RecapEducation {
  school: string;
  location: string;
  degree: string;
  fieldOfStudy: string;
  grade: string;
}

const recapEducationSchema = new mongoose.Schema({
  school: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  location: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  degree: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  fieldOfStudy: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  grade: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
});

export interface RecapEducationModel extends RecapBaseModel, RecapEducation {}

const RecapEducation = RecapBaseModel.discriminator("Education", recapEducationSchema);

export const RecapEducationModel = mongoose.model<RecapEducationModel>("Education");

// Accomplishments

export interface RecapAccomplishments {
  title: string;
  type: string;
}

const recapAccomplishmentsSchema = new mongoose.Schema({
  title: {
    type: String,
    required: true,
    maxlength: MAX_GENERAL_LENGTH,
  },
  type: {
    type: String,
    required: true,
    maxlength: MAX_GENERAL_LENGTH,
  },
});

export interface RecapAccomplishmentsModel extends RecapBaseModel, RecapAccomplishments {}

const RecapAccomplishments = RecapBaseModel.discriminator("Accomplishments", recapAccomplishmentsSchema);

export const RecapAccomplishmentsModel = mongoose.model<RecapAccomplishmentsModel>("Accomplishments");

// Skills

export interface RecapSkills {
  title: string;
}

const recapSkillsSchema = new mongoose.Schema({
  title: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
});

export interface RecapSkillsModel extends RecapBaseModel, RecapSkills {}

const RecapSkills = RecapBaseModel.discriminator("Skills", recapSkillsSchema);

export const RecapSkillsModel = mongoose.model<RecapSkillsModel>("Skills");

// Side Projects

export interface RecapSideProjects {
  title: string;
}

const recapSideProjectsSchema = new mongoose.Schema({
  title: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
});

export interface RecapSideProjectsModel extends RecapBaseModel, RecapSideProjects {}

const RecapSideProjects = RecapBaseModel.discriminator("SideProjects", recapSideProjectsSchema);

export const RecapSideProjectsModel = mongoose.model<RecapSideProjectsModel>("SideProjects");

// Organizations

export interface RecapOrganizations {
  organizationName: string;
  location: string;
  positions: string;
}

const recapOrganizationsSchema = new mongoose.Schema({
  organizationName: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  location: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  positions: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
});

export interface RecapOrganizationsModel extends RecapBaseModel, RecapOrganizations {}

const RecapOrganizations = RecapBaseModel.discriminator("Organizations", recapOrganizationsSchema);

export const RecapOrganizationsModel = mongoose.model<RecapOrganizationsModel>("Organizations");

// References

export interface RecapReferences {
  company: string;
  title: string;
  phoneNumber: string;
  email: string;
}

export const recapReferencesSchema = new mongoose.Schema({
  company: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  title: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  phoneNumber: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
  email: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
});

export interface RecapReferencesModel extends RecapBaseModel, RecapReferences {}

const RecapReferences = RecapBaseModel.discriminator("References", recapReferencesSchema);

export const RecapReferencesModel = mongoose.model<RecapReferencesModel>("References");

// Other

export interface RecapOther {
  title: string;
}

const recapOtherSchema = new mongoose.Schema({
  title: { type: String, required: true, maxlength: MAX_GENERAL_LENGTH },
});

export interface RecapOtherModel extends RecapBaseModel, RecapOther {}

const RecapOther = RecapBaseModel.discriminator("Other", recapOtherSchema);

export const RecapOtherModel = mongoose.model<RecapOtherModel>("Other");
