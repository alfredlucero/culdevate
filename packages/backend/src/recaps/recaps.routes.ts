import { Router } from "express";
import RecapsController from "./recaps.controller";
import authMiddleware from "../middleware/auth.middleware";
import validationMiddleware from "../middleware/validation.middleware";
import { RecapSchema } from "./recaps.validation";

const RecapsRoutes = {
  path: "/recaps",
  router: Router(),

  initializeRoutes() {
    // POST /recaps
    this.router.post("/", authMiddleware, validationMiddleware(RecapSchema, "body"), RecapsController.createRecap);

    // PATCH /recaps/:recapId
    this.router.patch(
      "/:recapId",
      authMiddleware,
      validationMiddleware(RecapSchema, "body"),
      RecapsController.updateRecap,
    );

    return this.router;
  },
};

export default RecapsRoutes;
