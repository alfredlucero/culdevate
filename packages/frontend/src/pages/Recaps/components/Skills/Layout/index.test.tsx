import React from "react";
import { render, fireEvent, waitForElementToBeRemoved } from "@testing-library/react";
import SkillsLayout, { SkillsLayoutProps } from "./index";
import { RecapSkills, RecapKind } from "../../../recaps.interface";
import * as RecapsService from "../../../recaps.service";
import { RecapFields } from "../../../recaps.schema";

const sampleRecapSkills: RecapSkills = {
  title: "Tagalog",
  kind: RecapKind.Skills,
  userId: "userId",
  _id: "skillsId",
  proficiency: "Intermediate",
  bulletPoints: [
    "Not a fluent speaker but can understand every day Tagalog speaking well",
    "Can speak some Tagalog here and there but struggle with grammar",
  ],
};
const defaultProps: SkillsLayoutProps = {
  recaps: [sampleRecapSkills],
  onGoBackToLanding: () => {},
  onCreateRecapSuccess: () => {},
  onUpdateRecapSuccess: () => {},
  onDeleteRecapSuccess: () => {},
};

describe("<SkillsLayout />", () => {
  test("should render without error", () => {
    const { container } = render(
      <SkillsLayout {...defaultProps} testId="skillsLayoutTestId" className="extra-skills-layout-class" />,
    );

    expect(container).toMatchSnapshot();
  });

  test("should show the empty state when there are no recaps", () => {
    const { getByTestId, queryByTestId } = render(<SkillsLayout {...defaultProps} recaps={[]} />);

    expect(queryByTestId("skillsRecap")).toBeNull();

    expect(getByTestId("skillsEmptyCard")).toBeVisible();
  });

  test("should show the recaps when there is data", () => {
    const { getByTestId, queryByTestId } = render(<SkillsLayout {...defaultProps} />);

    expect(queryByTestId("skillsEmptyCard")).toBeNull();

    expect(getByTestId("skillsRecap")).toBeVisible();
  });

  test("should show the success alert after deleting a recap successfully", async () => {
    const onDeleteRecapSuccessMock = jest.fn();
    const { getByText, queryByText, findByText } = render(
      <SkillsLayout {...defaultProps} onDeleteRecapSuccess={onDeleteRecapSuccessMock} />,
    );

    fireEvent.click(getByText("Delete"));

    expect(getByText("Are you sure you want to delete this?")).toBeVisible();

    jest.spyOn(RecapsService, "deleteRecap").mockImplementationOnce(() => Promise.resolve(sampleRecapSkills));

    fireEvent.click(getByText("Confirm Delete"));

    await waitForElementToBeRemoved(() => queryByText("Are you sure you want to delete this?"));

    expect(onDeleteRecapSuccessMock).toHaveBeenCalled();
    const successAlert = await findByText("You have successfully deleted the Skills Recap!");
    expect(successAlert).toBeVisible();
  });

  test("should show the error alert after failing to delete a recap", async () => {
    const onDeleteRecapSuccessMock = jest.fn();
    const { getByText, queryByText, findByText } = render(
      <SkillsLayout {...defaultProps} onDeleteRecapSuccess={onDeleteRecapSuccessMock} />,
    );

    fireEvent.click(getByText("Delete"));

    expect(getByText("Are you sure you want to delete this?")).toBeVisible();

    jest.spyOn(RecapsService, "deleteRecap").mockImplementationOnce(() => Promise.reject({}));

    fireEvent.click(getByText("Confirm Delete"));

    await waitForElementToBeRemoved(() => queryByText("Are you sure you want to delete this?"));

    expect(onDeleteRecapSuccessMock).not.toHaveBeenCalled();
    const errorAlert = await findByText("Something went wrong with deleting the Skills Recap!", {
      exact: false,
    });
    expect(errorAlert).toBeVisible();
  });

  test("should show the success alert after creating a recap successfully", async () => {
    const onCreateRecapSuccessMock = jest.fn();
    const { getByText, findByText, getByLabelText, getByTestId, queryByText } = render(
      <SkillsLayout {...defaultProps} onCreateRecapSuccess={onCreateRecapSuccessMock} />,
    );

    fireEvent.click(getByText("Add a Recap"));

    const createModal = await findByText("Create a Skills Recap");
    expect(createModal).toBeVisible();

    const recap: RecapSkills = {
      title: "Tagalog",
      kind: RecapKind.Skills,
      userId: "userId",
      _id: "skillsId",
      proficiency: "Intermediate",
      bulletPoints: [],
    };

    fireEvent.change(getByLabelText(RecapFields.skillsTitle), { target: { value: recap.title } });
    fireEvent.blur(getByLabelText(RecapFields.skillsTitle));

    fireEvent.change(getByLabelText(RecapFields.skillsProficiency), { target: { value: recap.proficiency } });
    fireEvent.blur(getByLabelText(RecapFields.skillsProficiency));

    expect(getByTestId("skillsSaveButton")).not.toBeDisabled();

    jest.spyOn(RecapsService, "createRecap").mockImplementationOnce(() => Promise.resolve(recap));

    fireEvent.click(getByTestId("skillsSaveButton"));

    await waitForElementToBeRemoved(() => queryByText("Create a Skills Recap"));

    expect(onCreateRecapSuccessMock).toHaveBeenCalled();
    expect(queryByText("You have successfully created a Skills Recap!")).toBeVisible();
  });

  test("should show the error alert after failing to create a recap", async () => {
    const onCreateRecapSuccessMock = jest.fn();
    const { getByText, findByText, getByLabelText, getByTestId } = render(
      <SkillsLayout {...defaultProps} onCreateRecapSuccess={onCreateRecapSuccessMock} />,
    );

    fireEvent.click(getByText("Add a Recap"));

    const createModal = await findByText("Create a Skills Recap");
    expect(createModal).toBeVisible();

    const recap: RecapSkills = {
      title: "Tagalog",
      kind: RecapKind.Skills,
      userId: "userId",
      _id: "skillsId",
      proficiency: "Intermediate",
      bulletPoints: [],
    };

    fireEvent.change(getByLabelText(RecapFields.skillsTitle), { target: { value: recap.title } });
    fireEvent.blur(getByLabelText(RecapFields.skillsTitle));

    fireEvent.change(getByLabelText(RecapFields.skillsProficiency), { target: { value: recap.proficiency } });
    fireEvent.blur(getByLabelText(RecapFields.skillsProficiency));

    expect(getByTestId("skillsSaveButton")).not.toBeDisabled();

    jest.spyOn(RecapsService, "createRecap").mockImplementationOnce(() => Promise.reject({}));

    fireEvent.click(getByTestId("skillsSaveButton"));

    const createErrorAlert = await findByText("Something went wrong with adding a Skills Recap!", {
      exact: false,
    });

    expect(createErrorAlert).toBeVisible();
    expect(getByText("Create a Skills Recap")).toBeVisible();
    expect(onCreateRecapSuccessMock).not.toHaveBeenCalled();
  });

  test("should show the success alert after updating a recap successfully", async () => {
    const onUpdateRecapSuccessMock = jest.fn();
    const { getByText, queryByText, findByText, getByTestId } = render(
      <SkillsLayout {...defaultProps} onUpdateRecapSuccess={onUpdateRecapSuccessMock} />,
    );

    fireEvent.click(getByText("Edit"));

    const editModal = await findByText("Edit your Skills Recap");
    expect(editModal).toBeVisible();

    expect(getByTestId("skillsSaveButton")).not.toBeDisabled();

    jest.spyOn(RecapsService, "updateRecap").mockImplementationOnce(() => Promise.resolve(sampleRecapSkills));

    fireEvent.click(getByTestId("skillsSaveButton"));

    await waitForElementToBeRemoved(() => queryByText("Edit your Skills Recap"));

    expect(queryByText("You have successfully updated a Skills Recap!")).toBeVisible();
    expect(onUpdateRecapSuccessMock).toHaveBeenCalled();
  });

  test("should show the error alert after failing to update a recap", async () => {
    const onUpdateRecapSuccessMock = jest.fn();
    const { getByText, queryByText, findByText, getByTestId } = render(
      <SkillsLayout {...defaultProps} onUpdateRecapSuccess={onUpdateRecapSuccessMock} />,
    );

    fireEvent.click(getByText("Edit"));

    const editModal = await findByText("Edit your Skills Recap");
    expect(editModal).toBeVisible();

    expect(getByTestId("skillsSaveButton")).not.toBeDisabled();

    jest.spyOn(RecapsService, "updateRecap").mockImplementationOnce(() => Promise.reject({}));

    fireEvent.click(getByTestId("skillsSaveButton"));

    const updateErrorAlert = await findByText("Something went wrong with updating a Skills Recap!", {
      exact: false,
    });
    expect(updateErrorAlert).toBeVisible();

    expect(queryByText("Edit your Skills Recap")).toBeVisible();
    expect(onUpdateRecapSuccessMock).not.toHaveBeenCalled();
  });
});
