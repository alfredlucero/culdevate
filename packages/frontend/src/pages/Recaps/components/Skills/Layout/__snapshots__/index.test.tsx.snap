// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SkillsLayout /> should render without error 1`] = `
<div>
  <div
    class="extra-skills-layout-class"
    data-testid="skillsLayoutTestId"
  >
    <header
      class="mb-8"
    >
      <button
        class="font-bold py-2 px-4 rounded bg-transparent text-teal-700 border border-teal-500 hover:bg-teal-500 hover:text-white hover:border-transparent mb-4"
        type="button"
      >
        <div
          class="flex items-center"
        >
          <div
            class="font-sans text-2xl mr-2"
            id=""
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-chevron-left fa-w-10 "
              data-icon="chevron-left"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
                fill="currentColor"
              />
            </svg>
          </div>
           Back to Recaps
        </div>
      </button>
      <div>
        <div
          class="flex justify-between w-full"
        >
          <h2
            class="font-mono font-bold text-gray-800 text-3xl "
          >
            Skills
          </h2>
          <button
            class="font-bold py-2 px-4 rounded font-bold py-2 px-4 rounded bg-teal-500 text-white hover:bg-teal-700 "
            type="button"
          >
            <div
              class="flex items-center"
            >
              Add a Recap
            </div>
          </button>
        </div>
        <p
          class="font-sans text-gray-800 text-base "
        >
          Recap the things you have learned and their proficiencies.
        </p>
      </div>
    </header>
    <section
      class=""
    >
      <div
        class="bg-white shadow-md rounded px-8 py-6 p-4 flex mb-4"
        data-testid="skillsRecap"
      >
        <div
          class="flex justify-center items-center p-3 rounded-lg bg-gray-200 recap-icon mr-4"
        >
          <div
            class="font-sans text-3xl text-teal-400"
            id=""
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-certificate fa-w-16 "
              data-icon="certificate"
              data-prefix="fas"
              focusable="false"
              role="img"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M458.622 255.92l45.985-45.005c13.708-12.977 7.316-36.039-10.664-40.339l-62.65-15.99 17.661-62.015c4.991-17.838-11.829-34.663-29.661-29.671l-61.994 17.667-15.984-62.671C337.085.197 313.765-6.276 300.99 7.228L256 53.57 211.011 7.229c-12.63-13.351-36.047-7.234-40.325 10.668l-15.984 62.671-61.995-17.667C74.87 57.907 58.056 74.738 63.046 92.572l17.661 62.015-62.65 15.99C.069 174.878-6.31 197.944 7.392 210.915l45.985 45.005-45.985 45.004c-13.708 12.977-7.316 36.039 10.664 40.339l62.65 15.99-17.661 62.015c-4.991 17.838 11.829 34.663 29.661 29.671l61.994-17.667 15.984 62.671c4.439 18.575 27.696 24.018 40.325 10.668L256 458.61l44.989 46.001c12.5 13.488 35.987 7.486 40.325-10.668l15.984-62.671 61.994 17.667c17.836 4.994 34.651-11.837 29.661-29.671l-17.661-62.015 62.65-15.99c17.987-4.302 24.366-27.367 10.664-40.339l-45.984-45.004z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
        <div
          class="flex-auto"
        >
          <h4
            class="font-mono font-bold text-gray-800 text-xl "
          >
            Tagalog
          </h4>
          <p
            class="font-sans text-gray-800 text-base "
          >
            Intermediate
          </p>
          <ul
            class="mt-3"
          >
            <li
              class="flex items-center"
            >
              <div
                class="font-sans text-2xl mr-2 text-teal-300"
                id=""
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-square fa-w-14 "
                  data-icon="square"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48z"
                    fill="currentColor"
                  />
                </svg>
              </div>
              <p
                class="font-sans text-gray-800 text-base pt-1"
              >
                Not a fluent speaker but can understand every day Tagalog speaking well
              </p>
            </li>
            <li
              class="flex items-center"
            >
              <div
                class="font-sans text-2xl mr-2 text-teal-300"
                id=""
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-square fa-w-14 "
                  data-icon="square"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48z"
                    fill="currentColor"
                  />
                </svg>
              </div>
              <p
                class="font-sans text-gray-800 text-base pt-1"
              >
                Can speak some Tagalog here and there but struggle with grammar
              </p>
            </li>
          </ul>
        </div>
        <div
          class="w-1/3 text-right"
        >
          <button
            class="font-bold py-2 px-4 rounded bg-transparent text-teal-700 border border-teal-500 hover:bg-teal-500 hover:text-white hover:border-transparent mr-2"
            id="skillsId"
            type="button"
          >
            <div
              class="flex items-center"
            >
              <div
                class="font-sans text-2xl mr-2"
                id=""
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-pencil-alt fa-w-16 "
                  data-icon="pencil-alt"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M497.9 142.1l-46.1 46.1c-4.7 4.7-12.3 4.7-17 0l-111-111c-4.7-4.7-4.7-12.3 0-17l46.1-46.1c18.7-18.7 49.1-18.7 67.9 0l60.1 60.1c18.8 18.7 18.8 49.1 0 67.9zM284.2 99.8L21.6 362.4.4 483.9c-2.9 16.4 11.4 30.6 27.8 27.8l121.5-21.3 262.6-262.6c4.7-4.7 4.7-12.3 0-17l-111-111c-4.8-4.7-12.4-4.7-17.1 0zM124.1 339.9c-5.5-5.5-5.5-14.3 0-19.8l154-154c5.5-5.5 14.3-5.5 19.8 0s5.5 14.3 0 19.8l-154 154c-5.5 5.5-14.3 5.5-19.8 0zM88 424h48v36.3l-64.5 11.3-31.1-31.1L51.7 376H88v48z"
                    fill="currentColor"
                  />
                </svg>
              </div>
               Edit
            </div>
          </button>
          <button
            class="font-bold py-2 px-4 rounded font-bold py-2 px-4 rounded bg-red-500 text-white hover:bg-red-700 "
            id="skillsId"
            type="button"
          >
            <div
              class="flex items-center"
            >
              <div
                class="font-sans text-2xl mr-2"
                id=""
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-trash-alt fa-w-14 "
                  data-icon="trash-alt"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
                    fill="currentColor"
                  />
                </svg>
              </div>
              Delete
            </div>
          </button>
        </div>
      </div>
    </section>
  </div>
</div>
`;
